specutils/spectra/spectrum1d.py

```diff
@@ -163,27 +182,51 @@ def __init__(self, flux=None, spectral_axis=None, wcs=None,
                                    "energy/wavelength/frequency equivalent.")
 
         # If flux and spectral axis are both specified, check that their lengths
-        # match or are off by one (implying the spectral axis stores bin edges)
+        # match or are off by one (implying the spectral axis stores bin edges).
+        # If we can't determine which flux axis corresponds to the spectral axis
+        # we raise an error.
         if flux is not None and spectral_axis is not None:
-            if spectral_axis.shape[0] == flux.shape[-1]:
+            if spectral_axis_index is None:
+                if flux.ndim == 1:
+                    self._spectral_axis_index = 0
+                else:
+                    matching_axes = []
+                    if bin_specification == "centers":
+                        add_elements = [0,]
+                    elif bin_specification == "edges":
+                        add_elements = [1,]
+                    elif bin_specification is None:
+                        add_elements = [0,1]
+                    for i in range(flux.ndim):
+                        for add_element in add_elements:
+                            if spectral_axis.shape[0] == flux.shape[i] + add_element:
+                                matching_axes.append(i)
+
+                    if len(matching_axes) == 1:
+                        self._spectral_axis_index = matching_axes[0]
+                    else:
+                        raise ValueError("Unable to determine which flux axis corresponds to "
+                                         "the spectral axis. Please specify spectral_axis_index"
+                                         " or provide a spectral_axis matching a flux axis.")
```

Comment:
While I like the general idea of "best guess" here, there's one failure mode that worries me: the case where there's a multi-d flux axis and one of those axes just by chance happens to match the spectral axis.  I can easily imagine someone writing code thinking the code just "knows" the correct axis, but then getting tripped up in a non-obvious way when there's an input that's ambiguous.

I'm not sure I have a clean solution though. A solution I don't love but *might* be good is to emit a warning whenever the guess is done, to hint to implementers they should really specify the axis if they can.  But that will probably lead to a lot of annoying warnings.

Maybe @keflavich or @Cadair have some idea how common this failure mode might be?  In optical-land I think it's rare enough it's not a big deal (since cubes are almost always a lot smaller than the spectral dimension) but I'm not as sure about other domains...

Reply:
Isn't it _required_ that at least one flux dimension to match the size of the spectral axis? I'm not sure it makes any sense to create a `Spectrum1D` with flux shape, say, `[10,10,200]` and provide a spectral axis of length 240. This error would get thrown in that case, and it would also be raised if one of the spatial dimensions coincidentally happened to also match the length of the spectral axis.

Reply:
Even in optical land, you'll work with data subsets, and these can have any shape.

If there's exactly _one_ match, you're OK.  If there are multiple matches or zero matches, error, right?

@rosteen What if you had x,y,spectral = [10,20,200] and a spectral axis was passed in with length 20?  That would not result in this error but would be wrong.  I'm .... not sure why a user would ever do this, though?

Reply:
Hm, I see the point now. I'm also not sure why a user would do that, but then again...there have been many times when I wasn't sure why a user did something 😆. I guess the question is "is this likely enough to justify the annoyance of either displaying a warning, or always requiring `spectral_axis_index`?" And I'm not sure of the answer.

Reply:
I'm loosely inclined toward no warning, but let's be prepared to post this conversation as a response to the first user error that results in this corner case.